### PR TITLE
feat: Implement app layer Import

### LINKED ISSUE #6: Implement Data Manager - Import recorded data
Per ADR: https://docs.edgexfoundry.org/3.0/design/adr/application/0025-Record-and-Replay/#upload-endpoint-import

### PR DESCRIPTION
**Pull request by lenny-intel**
closes #6

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-record-replay/blob/main/.github/CONTRIBUTING.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?) **TBD**
  <link to docs PR>

## Testing Instructions

From compose builder run `make run no-secty ds-virtual` 
Build app service
run `WRITABLE_LOGLEVEL=DEBUG ./app-record-replay -cp -d -o | grep "ARR "` to run app service 
Send POST the following to `localhost:59712/api/v3/record` to record data
{
    "duration" : 60000000000,
    "eventLimit" : 10
}
After less than 1min logs shows :
```
ARR Process Recorded Data: 10 events in Zs have been saved for replay
```
Run `curl localhost:59712/api/v3/data -o test.json`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T22:51:46.075333336Z app=app-record-replay source=manager.go:465 msg="ARR Export: Loaded 3 devices for export"
level=DEBUG ts=2023-08-28T22:51:46.080144964Z app=app-record-replay source=manager.go:482 msg="ARR Export: Loaded 3 devices profiles for export"     
level=DEBUG ts=2023-08-28T22:51:46.080179064Z app=app-record-replay source=manager.go:485 msg="ARR Export: Exporting 10 events, 3 devices and 3 device profiles"
level=DEBUG ts=2023-08-28T22:51:46.080191064Z app=app-record-replay source=controller.go:244 msg="ARR Export - Exporting as JSON w/o compression"
```
run `curl localhost:59712/api/v3/data?compression=gzip -o test.gz`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T22:53:46.282936979Z app=app-record-replay source=manager.go:485 msg="ARR Export: Exporting 10 events, 3 devices and 3 device profiles"
level=DEBUG ts=2023-08-28T22:53:46.282986479Z app=app-record-replay source=controller.go:269 msg="ARR Export - Exporting as JSON using GZIP compression"
```
run `curl localhost:59712/api/v3/data?compression=zlib -o test.zlib`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T22:55:36.990875904Z app=app-record-replay source=manager.go:485 msg="ARR Export: Exporting 10 events, 3 devices and 3 device profiles"
level=DEBUG ts=2023-08-28T22:55:36.990915004Z app=app-record-replay source=controller.go:256 msg="ARR Export - Exporting as JSON using ZLIB compression"
```
run `curl -d @./test.json  localhost:59712/api/v3/data -H "Content-Type: application/json"`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T22:58:29.269147886Z app=app-record-replay source=controller.go:320 msg="ARR Import - Importing as JSON w/o compression"    
level=DEBUG ts=2023-08-28T22:58:29.272133214Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Integer-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T22:58:29.273312025Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-UnsignedInteger-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T22:58:29.274273234Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Boolean-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T22:58:29.278831577Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-UnsignedInteger-Device"
level=DEBUG ts=2023-08-28T22:58:29.284153527Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Boolean-Device"
level=DEBUG ts=2023-08-28T22:58:29.289430277Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Integer-Device"
level=DEBUG ts=2023-08-28T22:58:29.289491077Z app=app-record-replay source=manager.go:543 msg="ARR Import: Imported 10 events, 3 devices and 3 device profiles"
```
run `curl --data-binary @./test.gz  localhost:59712/api/v3/data -H "Content-Type: application/json" -H "Content-Encoding: gzip"`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T23:03:36.252922385Z app=app-record-replay source=controller.go:324 msg="ARR Import - Importing as JSON using GZIP compression"
level=DEBUG ts=2023-08-28T23:03:36.257091323Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Boolean-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T23:03:36.258694937Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Integer-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T23:03:36.260214251Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-UnsignedInteger-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T23:03:36.265608599Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Integer-Device"
level=DEBUG ts=2023-08-28T23:03:36.270789146Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-UnsignedInteger-Device"
level=DEBUG ts=2023-08-28T23:03:36.275482288Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Boolean-Device"
level=DEBUG ts=2023-08-28T23:03:36.275553889Z app=app-record-replay source=manager.go:543 msg="ARR Import: Imported 10 events, 3 devices and 3 device profiles"
```
run `curl --data-binary @./test.zlib  localhost:59712/api/v3/data -H "Content-Type: application/json" -H "Content-Encoding: deflate"`
Verify logs contain:
```
level=DEBUG ts=2023-08-28T23:05:23.862386936Z app=app-record-replay source=controller.go:333 msg="ARR Import - Importing as JSON using ZLIB compression"
level=DEBUG ts=2023-08-28T23:05:23.866583049Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Boolean-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T23:05:23.868427154Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-Integer-Device not updated: Can not update existing device profiles"level=DEBUG ts=2023-08-28T23:05:23.87018926Z app=app-record-replay source=manager.go:624 msg="ARR Import: Existing device profile Random-UnsignedInteger-Device not updated: Can not update existing device profiles"
level=DEBUG ts=2023-08-28T23:05:23.875848277Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Boolean-Device"
level=DEBUG ts=2023-08-28T23:05:23.881492094Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-Integer-Device"
level=DEBUG ts=2023-08-28T23:05:23.887376812Z app=app-record-replay source=manager.go:591 msg="ARR Import: Updated existing device Random-UnsignedInteger-Device"
level=DEBUG ts=2023-08-28T23:05:23.887476613Z app=app-record-replay source=manager.go:543 msg="ARR Import: Imported 10 events, 3 devices and 3 device profiles"
```
run `curl -d @./test.json  localhost:59712/api/v3/data?overwire=false -H "Content-Type: application/json"`
Verify logs **JUST** contain (i.e no messages about Devices or Profiles:
```
level=DEBUG ts=2023-08-28T23:07:57.325709862Z app=app-record-replay source=controller.go:320 msg="ARR Import - Importing as JSON w/o compression"    
level=DEBUG ts=2023-08-28T23:07:57.334123988Z app=app-record-replay source=manager.go:543 msg="ARR Import: Imported 10 events, 3 devices and 3 device profiles"
```
From EdgeX UI remove all the Devices and then remove all the Device Profiles
run `curl -d @./test.json  localhost:59712/api/v3/data?overwire=false -H "Content-Type: application/json"`
Verify logs contain add messages for  Devices or Profiles:
```
level=DEBUG ts=2023-08-28T23:22:37.943454305Z app=app-record-replay source=controller.go:320 msg="ARR Import - Importing as JSON w/o compression"
level=DEBUG ts=2023-08-28T23:22:37.950126946Z app=app-record-replay source=manager.go:614 msg="ARR Import: Add new device profile Random-Integer-Device"
level=DEBUG ts=2023-08-28T23:22:37.953212966Z app=app-record-replay source=manager.go:614 msg="ARR Import: Add new device profile Random-UnsignedInteger-Device"
level=DEBUG ts=2023-08-28T23:22:37.955899182Z app=app-record-replay source=manager.go:614 msg="ARR Import: Add new device profile Random-Boolean-Device"
level=DEBUG ts=2023-08-28T23:22:37.960826412Z app=app-record-replay source=manager.go:563 msg="ARR Import: Added new device Random-UnsignedInteger-Device"
level=DEBUG ts=2023-08-28T23:22:37.96527644Z app=app-record-replay source=manager.go:563 msg="ARR Import: Added new device Random-Boolean-Device"
level=DEBUG ts=2023-08-28T23:22:37.969019363Z app=app-record-replay source=manager.go:563 msg="ARR Import: Added new device Random-Integer-Device"
level=DEBUG ts=2023-08-28T23:22:37.969071563Z app=app-record-replay source=manager.go:543 msg="ARR Import: Imported 10 events, 3 devices and 3 device profiles"
```

### CODE REVIEW

File: internal/application/manager.go [hunk 1]
```diff
@@ -521,8 +512,119 @@ func (m *dataManager) loadDevices() error {
...
+		updateRequest := requests.NewUpdateDeviceRequest(updateDevice)
+		_, err = deviceClient.Update(context.Background(), []requests.UpdateDeviceRequest{updateRequest})
+		if err != nil {
+			return fmt.Errorf("failed to update device %s in system: %w", device.Name, err)
+		}
+
+		m.appSvc.LoggingClient().Debugf("ARR Import: Updated existing device %s", device.Name)
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/application/manager.go:594]: Nope, they are independent of each other.

File: internal/application/manager.go [hunk 2]
```diff
@@ -521,8 +512,119 @@ func (m *dataManager) loadDevices() error {
...
+}
+
+func (m *dataManager) uploadProfiles(profiles []coreDtos.DeviceProfile, overwrite bool) error {
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/application/manager.go:597]: Nope, they are independent of each other.

File: internal/application/manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply internal/application/manager.go:594]: gotcha

**Review by jim-wang-intel — APPROVED** [review]
LGTM

File: internal/application/manager.go [hunk 3]
```diff
@@ -521,8 +512,119 @@ func (m *dataManager) loadDevices() error {
...
+	for _, profile := range profiles {
+		_, err := profileClient.DeviceProfileByName(context.Background(), profile.Name)
+		if err != nil && err.Code() != http.StatusNotFound {
+			return fmt.Errorf("failed check if profile %s exists in system: %w", profile.Name, err)
+		}
+
+		addRequest := requests.NewDeviceProfileRequest(profile)
+
+		// If got here with an err then the Device Profile doesn't exist in the system, so must add it.
+		if err != nil {
```
Comment (brian-intel) [inline internal/application/manager.go:608]: should this use the same logic as the uploadDevices()?

err != nil && err.Code() == http.StatusNotFound

Comment (lenny-intel) [reply internal/application/manager.go:608]: Good question. I'll have to look at the API to see what it returns in this the Not Found case vs. general error. 
I think I did look, but will verify.

Comment (lenny-intel) [reply internal/application/manager.go:608]: @brian-intel,  Yes it does return 404. Great catch. I'll submit a fix PR later today. THX!

**Pull request merged**